### ISSUE TITLE
[Training]  Integrate a new Post Training quantization method

### ISSUE THREAD

**Issue by LucasFischer123** [microsoft/onnxruntime#14594]
### Describe the issue

Hi,

I am interested in developing a custom post-training quantization method using ONNXRuntime, and I require access to the gradients of each layer after processing a batch of images. Ideally, I would like to be able to do this with any ONNX model without having to create a custom training graph each time. The gradient I need to access does not require calculation of any loss.

Based on my understanding, I believe I need to explore the training API and ONNXBlock, but if I am correct, I may need to reconstruct the training graph which goes against what I am looking to do.

I have reviewed the relevant resources at the following links:
https://github.com/microsoft/onnxruntime/tree/main/orttraining/orttraining/python/training/api
https://github.com/microsoft/onnxruntime/blob/main/orttraining/orttraining/python/training/onnxblock/README.md

Does anyone have any suggestions or ideas regarding this matter?
Is it mandatory to use pytorch ?

Thank you very much.

### To reproduce

Not applicable

### Urgency

Working on that right now, I would be glad to have some guidelines

### ONNX Runtime Installation

Built from Source

### ONNX Runtime Version or Commit ID

0c6037b5abe571fc43a55ef7a9d2f846820fbe5d

### PyTorch Version

latest

### Execution Provider

Default CPU

**Comment by baijumeswani** [comment]
Hi @LucasFischer123. 

I am not very clear on the question here. If you don't mind, would you please elaborate? My question is:
1. Are you looking to perform training of any kind (including quantization aware training) using onnxruntime?
    - for this, we are currently developing a solution and you should use onnxblock and the training apis once the solution is ready.
2. Or are you looking to only do post training quantization (using either static or dynamic quantization)?
    - For this, you should use the onnxruntime quantization tool. More details at https://onnxruntime.ai/docs/performance/quantization.html
3. Or are you looking to build a custom gradient graph based on your onnx model?
    - For this, you can explore using the `GradientGraphBuilder` in onnxruntime: https://github.com/microsoft/onnxruntime/blob/main/orttraining/orttraining/python/orttraining_pybind_state.cc#L800. Let me know if this is what you want, I can provide more examples on how to use this utility.
4. Or something else?

**Comment by LucasFischer123** [comment]
Hi @baijumeswani 

Thanks a lot for the quick answer.

What we are trying to do is closely related to 1). 

We aren't doing quantization aware training but we need to acquire the gradient at each layer for some kind of post training quantization. So in a nutshell, we need to have some kind of backward propagation and get the gradient. Is it clearer for you ? 

Couple questions, 

1. can we start using onnxblock and the training api right now ? 
2. If yes, do you have tutorials or examples that we could use or help you build those blocks ? 
3. With the ONNXblock and training api, is it build to be able to work on any kind of neural networks with convolution.
4. Will it work for transformer

Thanks a lot for your help!

**Comment by baijumeswani** [comment]
1. Yes, you can use onnxblock and the training api. It should be ok to use now.
2. yes, we have a couple of simple examples:
    - https://github.com/microsoft/onnxruntime-training-examples/tree/master/on_device_training
 3. If you encounter something failing, please reach out through github issues. We will address them at the earliest possibility.
 4. Yes, it should work for transformer related models.

**Comment by LucasFischer123** [comment]
Thanks @baijumeswani, super clear. We will start using that !
I am closing the issue in the meantime !
